### PR TITLE
Allow storages to alias even when the deleter is deleteNothing

### PR DESCRIPTION
**Pull request by Anstow**
This improves the debugging experience for devices that provide custom storage pointer.

### CODE REVIEW

**Comment by colesbury** [comment]
@kurtamohler, can you please look at this?

**Comment by kurtamohler** [comment]
I'd like to understand more of the context for this why this change is needed. Have you considered making your custom DataPtr use `refcounted_deleter` (by calling the `maybeApplyRefcountedDeleter` function in `c10/core/RefcountedDeleter.h`), and is there a reason why that wouldn't work in your case?

**Comment by Anstow** [comment]
Like XLA we don't really have a concept of storage on our device that matches up with the pytorch definition of storage. So we're faking it by passing a pointer to an internal class with a empty destructor (we manage the lifetime of said class internally). XLA passes a `nullptr` rather than a real ptr but we need to pass a real ptr so we can handle `set_.source_Storage` correctly. I think it would technically be possible to let pytorch handle the memory management of our internal class but it would require significant refactoring on our end and may make our view handling code worse

References in XLA
Note on XLA storage handling here: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.h#L334
Where the XLA storage is actually initialized: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L144

File: c10/core/Storage.cpp [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
 
 bool isSharedStorageAlias(const Storage& storage0, const Storage& storage1) {
-  c10::DeleterFnPtr deleter_expected = &c10::refcounted_deleter;
   c10::DeleterFnPtr deleter0 = storage0.data_ptr().get_deleter();
   c10::DeleterFnPtr deleter1 = storage1.data_ptr().get_deleter();
 
-  if ((deleter0 != deleter_expected) || (deleter1 != deleter_expected)) {
+  if (deleter0 != deleter1 ||
+      (deleter0 != &c10::refcounted_deleter &&
+       deleter0 != &c10::detail::deleteNothing)) {
```
Comment (kurtamohler) [inline c10/core/Storage.cpp:13]: I wonder if we should avoid changing what happens with non-XLA storages for now

```suggestion
  if ((deleter0 != deleter_expected) || (deleter1 != deleter_expected)) {
    return false;
  } else if (storage0.device_type() == c10::kXLA && storage1.device_type() == c10::kXLA && (deleter0 != deleter1 ||
      (deleter0 != &c10::refcounted_deleter &&
       deleter0 != &c10::detail::deleteNothing))) {
```

**Pull request closed without merging**